### DISCUSSION TITLE
Controlling resistance with inclination on elliptical

### DISCUSSION THREAD

**Discussion by realericmc** [cagnulein/qdomyos-zwift#2306] [Q&A]
Is there any way to automatically control the resistance on an elliptical based on the inclination?  I have inclination control working with zwift, which is cool, but the effect on hills is not very realistic.  The track tilts up so the step changes to something closer to a climb, but with no resistance change I end up spinning in tall fast circles instead of it getting harder as would happen on a real incline.  I dug through the settings but couldn't find an automatic resistance control.  thanks!

**Comment by cagnulein** [comment]
are you able to change the resistance from the resistance tile on qz
manually? if so yes i can add it

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 24 apr 2024 alle 20:48 Eric McMurry ***@***.***>
ha scritto:

**Comment by realericmc** [comment]
yep, that works fine.  thanks!

**Comment by cagnulein** [comment]
so you would prefere resistance changes instead of inclination?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 24 apr 2024 alle 20:52 Eric McMurry ***@***.***>
ha scritto:

**Comment by realericmc** [comment]
both would be nice.  I was thinking in terms of having the resistance change based on the inclination, along with an offset and gain so the resistance relative to the inclination could be adjusted.

**Comment by cagnulein** [comment]
ok i will do it

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 24 apr 2024 alle 20:59 Eric McMurry ***@***.***>
ha scritto:

**Comment by realericmc** [comment]
cool!  thanks Roberto!  more coffee incoming

**Comment by cagnulein** [comment]
done! it will be in the today version.
the zwift resistance offset and zwift resistance gain will help you to calibrate the resistance response

**Comment by realericmc** [comment]
Nice!  Thanks!

**Reply by realericmc** [reply]
well, I couldn't find the settings to turn it on or adjust it.  I tried digging through all of them but it escaped me.  Where are those settings?  thanks!

**Reply by cagnulein** [reply]
under bike options

![image](https://github.com/cagnulein/qdomyos-zwift/assets/48216/471d82cf-64ef-4095-a13c-cb1ed12540e8)

**Reply by realericmc** [reply]
ah, the one spot I didn't look :-).  so, I turn on erg mode and then use the resistance offset and gain to adjust?

**Comment by cagnulein** [comment]
nope don't turn on erg mode, just use as usual . resistance will change
according to the inclination. you can increase the offset and the gain to
adjust the response

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 25 apr 2024 alle 16:38 Eric McMurry ***@***.***>
ha scritto:

**Reply by realericmc** [reply]
got it.  I will try it.  so, to turn it off you set the gain to 0?

**Comment by cagnulein** [comment]
yes

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 25 apr 2024 alle 16:41 Eric McMurry ***@***.***>
ha scritto:

**Reply by realericmc** [reply]
it works, after a fashion.  It seems to have trouble with negative slopes.  For a while it seemed to be working right, then it started behaving a bit randomly with the inclination going too high while the resistance was lower (like they flipped) and the inclination also got stuck above one when the zwift slope went down (I've seen this part happen before)
[debug-Thu_Apr_25_09_34_45_2024.log.zip](https://github.com/cagnulein/qdomyos-zwift/files/15112007/debug-Thu_Apr_25_09_34_45_2024.log.zip)

**Comment by cagnulein** [comment]
does your elliptical have negative inclination? if so which is negative
limit?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 25 apr 2024 alle 17:00 Eric McMurry ***@***.***>
ha scritto:

**Reply by realericmc** [reply]
it doesn't.  it's inclination goes from 1 to 20, which is actually 1.5 to 15 degrees of slope.  zwift does do negative though and that is where I saw the called for resistance go up to 16-20

**Comment by cagnulein** [comment]
ok i will check the log tomorrow and i will let you know!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 25 apr 2024 alle 17:39 Eric McMurry ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
got the issue! it's an easy one! i will send you a new one!

**Reply by realericmc** [reply]
that worked a lot better!

btw, the inclination is still getting stuck when zwift switches to negative slope.  There are a couple of instances in this log.  in the last one it stuck at 8, while the slope was negative.  the resistance went down though.
Thanks!
[debug-Thu_Apr_25_11_56_16_2024.log.zip](https://github.com/cagnulein/qdomyos-zwift/files/15113567/debug-Thu_Apr_25_11_56_16_2024.log.zip)

**Comment by cagnulein** [comment]
Stupid me, I have to do the same thing for the inclination as well

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 25 apr 2024 alle 19:22 Eric McMurry ***@***.***>
ha scritto:

**Reply by realericmc** [reply]
well, unfortunately the incline is still sticking when the slope goes negative.  There should be a couple of examples in this.  thanks!

[debug-Thu_Apr_25_19_40_32_2024.log.zip](https://github.com/cagnulein/qdomyos-zwift/files/15123738/debug-Thu_Apr_25_19_40_32_2024.log.zip)

**Reply by cagnulein** [reply]
hi @realericmc actually in this log I saw only positive inclination, the negative one are filtered out sending to the elliptical the value "0". Is it what you are seeing?

**Reply by realericmc** [reply]
ah, maybe that is the issue.  the elliptical only goes down to 1, so perhaps it is filtering out the zero?

**Reply by cagnulein** [reply]
are you talking about resistance or inclination?

**Reply by realericmc** [reply]
inclination.  actually both the resistance and inclination on this are 1-20, although the resistance is not showing the same issue.

**Reply by cagnulein** [reply]
hah ok, yes for the resistance i set the min as 1 for the inclination I set it to 0. so that's the issue. let me set to 1 as well. i will send you this later today

**Reply by realericmc** [reply]
no rush.  thanks for fixing it!  now if I could only get zwift to actually report the inclines so I could get credit :-)

**Comment by cagnulein** [comment]
eheheh you have to contact zwift for this :)

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle ore 14:37 Eric McMurry <
***@***.***> ha scritto:

**Reply by realericmc** [reply]
I've actually seen mention that some treadmills will report this to zwift and it will throw it in the fit files.  I didn't see any details though so I don't know what the mechanism is.

**Comment by cagnulein** [comment]
yes with QZ i can see the inclination on zwift in the strava workout. Only
there. You have to use the wahoo tread device in zwift for this

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle ore 14:40 Eric McMurry <
***@***.***> ha scritto:

**Reply by realericmc** [reply]
do you think it would be possible to mimic that to zwift by passing it back its own inclination in the format of a wahoo tread?

**Comment by cagnulein** [comment]
i'm already doing for you

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle ore 14:44 Eric McMurry <
***@***.***> ha scritto:

**Reply by realericmc** [reply]
heh, you're faster than I can type

**Comment by cagnulein** [comment]
i mean it's already in the version that you already have

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle ore 14:47 Eric McMurry <
***@***.***> ha scritto:

**Reply by realericmc** [reply]
ah, gotcha.  how do I turn that on?

**Comment by cagnulein** [comment]
It's on by default. Did you use the wahoo tread or the phone name on zwift?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle 15:29 Eric McMurry ***@***.***>
ha scritto:

**Comment by realericmc** [comment]
I use the wahoo tread

**Comment by cagnulein** [comment]
So it should works. You can ask about this on the qz Facebook group also.
There are a lot of treadmill zwift users

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle 16:14 Eric McMurry ***@***.***>
ha scritto:

**Reply by realericmc** [reply]
got it.  so it should be sending that data back to zwift even if it is an elliptical?

**Comment by cagnulein** [comment]
Yes because qz is sending them as a treadmill to zwift

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle 16:36 Eric McMurry ***@***.***>
ha scritto:

**Reply by realericmc** [reply]
okay, not sure what else to try then.  not the end of the world to not have that data.  thanks!

**Comment by cagnulein** [comment]
Did you check on the strava session? Do you have a link of one?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle 16:42 Eric McMurry ***@***.***>
ha scritto:

**Comment by realericmc** [comment]
yep, they have 0 elevation in them.  Here's the fit from the last one that had the crash during it:
[Zwift_2019_Worlds_Harrogate_Circuit_in_Yorkshire.fit.zip](https://github.com/cagnulein/qdomyos-zwift/files/15132016/Zwift_2019_Worlds_Harrogate_Circuit_in_Yorkshire.fit.zip)

**Reply by cagnulein** [reply]
check this for example https://www.strava.com/activities/11063028798# i can see the elevation from zwift on my qz setup

**Reply by realericmc** [reply]
cool, that looks like what I was hoping for.  My runs just show 0 elevation though.  Those get synced over to Garmin connect also and it shows zero too.

**Comment by cagnulein** [comment]
maybe zwift patched this (i'm not using zwift anymore, just qz). So that's
why i said to ask to the FB group, they will give you an actual answer for
sure ;)

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 apr 2024 alle ore 16:53 Eric McMurry <
***@***.***> ha scritto:

**Reply by realericmc** [reply]
Don't know why I didn't look before, but when I checked a recorded workout in the qz app, I also don't see any elevation, so I guess it makes sense that it wouldn't show up back in zwift either.  Do I need to turn something on in the app for that, or is it just an artifact of zwift not sending anything?

**Reply by cagnulein** [reply]
but do you see the inclination tile with the right value on qz?

**Comment by realericmc** [comment]
Yep, inclination is there and working, and I see a chart for that.  Should I also see one for elevation?

**Comment by cagnulein** [comment]
you should see the right value in the elevation gain tile too . did you
enable it?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 6 mag 2024 alle 19:49 Eric McMurry ***@***.***>
ha scritto:

> Yep, inclination is there and working, and I see a chart for that. Should
> I also see one for elevation?
>

**Comment by realericmc** [comment]
Yep, that is on, and I see it incrementing.  Digging a bit deeper, I do see increasing elevation in the fit file, I just didn’t see a chart on the app so wasn’t sure.  It only shows elevation increase though, and with a steady incline.  I’m guessing from that data that it is doing that based on the incline of the elliptical rather than the slope coming from zwift?  For this elliptical it only goes down to 1 and the numbers are not slope so it doesn’t translate directly to elevation.

**Reply by cagnulein** [reply]
there isn't a elevation chart in qz, never added :) 

but if you see in the fit file qz is doing the right thing :) 

try to ask about this on the qz facebook group

**Comment by realericmc** [comment]
No worries, I just noticed it wasn’t there, wasn’t asking for it.  If data is in fit, presumably it is sent on.  Sorry, I don’t have fb, so I will drop the request and hope that the zwift folks fix this at some point.  Thanks!

**Reply by cagnulein** [reply]
maybe you can just use the fit file from qz instead of the zwift one :)

**Comment by realericmc** [comment]
Yeah, that one is better.  Lots of manual shuffling to do that though.  Also, the elevation data in there is still wrong since it is a recording the inclination of the elliptical, which does not match the elevation from zwift (well, it’s all guesswork anyways I suppose since you can’t directly mimic hills).  Thanks!

**Reply by cagnulein** [reply]
qz can mimic hills, just set the zwift credentials in qz and qz will do the rest!

**Reply by cagnulein** [reply]
what do you mean it doesn't match. it must be: when zwift shows 3 the elliptical should be forced to 3. what do you see instead?

about the elevation gain, that's right, elevation gain goes only one way, it counts only the positive gain, not the negative one

**Comment by realericmc** [comment]
Yep, that is working well.  I just meant that it was a simulation since it is an elliptical, so the numbers around speed and elevation are estimations.  The only issue I have in the data is the elevation increase which only goes up and always goes up.  Thanks!

**Comment by realericmc** [comment]
This elliptical only goes down to 1 for incline, and the incline number is not an angle, just a unit-less number from 1 to 20 (actual angles I measured as 1.5-15º).  So, recording that shows a constant increase even when the slope is flat or negative as requested by zwift.  

The elevation gain is shown as elevation in the fit, so if I pull that up on Garmin connect for instance, it just looks like I was climbing a hill the whole time.  It is swift’s problem really, since they could just put that into the data like they do for cycling.

**Comment by cagnulein** [comment]
hah ok if you want I can remap the inclination value to a table so you can
assign a real inclination value
let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 6 mag 2024 alle 23:10 Eric McMurry ***@***.***>
ha scritto:

> This elliptical only goes down to 1 for incline, and the incline number is
> not an angle, just a unit-less number from 1 to 20 (actual angles I
> measured as 1.5-15º). So, recording that shows a constant increase even
> when the slope is flat or negative as requested by zwift.
>
> The elevation gain is shown as elevation in the fit, so if I pull that up
> on Garmin connect for instance, it just looks like I was climbing a hill
> the whole time. It is swift’s problem really, since they could just put
> that into the data like they do for cycling.
>
